core/src/main/java/com/datastrato/gravitino/storage/relational/service/TagMetaService.java

```diff
@@ -0,0 +1,191 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datastrato.gravitino.storage.relational.service;
+
+import com.datastrato.gravitino.Entity;
+import com.datastrato.gravitino.HasIdentifier;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.exceptions.NoSuchEntityException;
+import com.datastrato.gravitino.meta.TagEntity;
+import com.datastrato.gravitino.storage.relational.mapper.TagMetaMapper;
+import com.datastrato.gravitino.storage.relational.mapper.TagMetadataObjectRelMapper;
+import com.datastrato.gravitino.storage.relational.po.TagPO;
+import com.datastrato.gravitino.storage.relational.utils.ExceptionUtils;
+import com.datastrato.gravitino.storage.relational.utils.POConverters;
+import com.datastrato.gravitino.storage.relational.utils.SessionUtils;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class TagMetaService {
+
+  private static final TagMetaService INSTANCE = new TagMetaService();
+
+  public static TagMetaService getInstance() {
+    return INSTANCE;
+  }
+
+  private TagMetaService() {}
+
+  public List<TagEntity> listTagsByNamespace(Namespace ns) {
+    String metalakeName = ns.level(0);
+    List<TagPO> tagPOs =
+        SessionUtils.getWithoutCommit(
+            TagMetaMapper.class, mapper -> mapper.listTagPOsByMetalake(metalakeName));
+    return tagPOs.stream()
+        .map(tagPO -> POConverters.fromTagPO(tagPO, ns))
+        .collect(Collectors.toList());
+  }
+
+  public TagEntity getTagByIdentifier(NameIdentifier ident) {
+    String metalakeName = ident.namespace().level(0);
+    TagPO tagPO = getTagPOByMetalakeAndName(metalakeName, ident.name());
+    return POConverters.fromTagPO(tagPO, ident.namespace());
+  }
+
+  public void insertTag(TagEntity tagEntity, boolean overwritten) throws IOException {
+    Namespace ns = tagEntity.namespace();
+    String metalakeName = ns.level(0);
+    AtomicReference<Long> metalakeId = new AtomicReference<>(null);
+
+    try {
+      SessionUtils.doMultipleWithCommit(
+          () -> metalakeId.set(MetalakeMetaService.getInstance().getMetalakeIdByName(metalakeName)),
```

Comment:
I was thinking that `select...` cannot block the `delete` operation, so we need to have a metalake lock here to avoid  user deleting the metalake while we're querying it.

Reply:
It seems to we need use `select xxx for update` to obtain the write lock here, I'm not very it will cause dead lock or other problem or not.

Reply:
One choice is to use `select...for update`.

Reply:
I was planning to avoid using tag lock or tree lock, instead using transactions to keep consistency, but seems it is not achievable, still we have to introduce lock here.

Reply:
According to the current design, I think we'd better add locks in the upper layer to avoid duplication of internal locks and database locks.

Reply:
Since there is a upper lock mechanism like tree lock, using it is a reasonable solution.